tracer/context.go

```diff
@@ -2,41 +2,35 @@ package tracer
 
 import (
 	"context"
+
+	"github.com/DataDog/dd-trace-go/dd"
 )
 
-var spanKey = "datadog_trace_span"
+type contextKey struct{}
 
-// ContextWithSpan will return a new context that includes the given span.
-// DEPRECATED: use span.Context(ctx) instead.
-func ContextWithSpan(ctx context.Context, span *Span) context.Context {
-	if span == nil {
-		return ctx
-	}
-	return span.Context(ctx)
-}
+var activeSpanKey = contextKey{}
 
-// SpanFromContext returns the stored *Span from the Context if it's available.
-// This helper returns also the ok value that is true if the span is present.
-func SpanFromContext(ctx context.Context) (*Span, bool) {
-	if ctx == nil {
-		return nil, false
-	}
-	span, ok := ctx.Value(spanKey).(*Span)
-	return span, ok
+// ContextWithSpan returns a copy of the given context which includes the span.
+func ContextWithSpan(ctx context.Context, s Span) context.Context {
+	return context.WithValue(ctx, activeSpanKey, s)
 }
 
-// SpanFromContextDefault returns the stored *Span from the Context. If not, it
-// will return an empty span that will do nothing.
-func SpanFromContextDefault(ctx context.Context) *Span {
-
-	// FIXME[matt] is it better to return a singleton empty span?
-	if ctx == nil {
-		return &Span{}
+// SpanFromContext returns the span contained in the given context. If no span is
+// found, it returns nil.
+func SpanFromContext(ctx context.Context) Span {
+	v := ctx.Value(activeSpanKey)
+	if s, ok := v.(dd.Span); ok {
+		return s
 	}
+	return nil
+}
 
-	span, ok := SpanFromContext(ctx)
-	if !ok {
-		return &Span{}
+// StartSpanFromContext returns a new span with the given operation name and options. If a span
+// is found in the context, it will be used as the parent of the resulting span.
+func StartSpanFromContext(ctx context.Context, operationName string, opts ...StartSpanOption) (Span, context.Context) {
+	if s := SpanFromContext(ctx); s != nil {
+		opts = append(opts, ChildOf(s.Context()))
```

Comment:
Nice catch. I have also thought of this. Since we iterate through all options, I guess in this case (as you can see in [the code](https://github.com/DataDog/dd-trace-go/blob/gbbr/ot/tracer/tracer.go#L211-L213)) the last one will override any previous ones. So the one in the context will take precedence over the user's argument, which is what you would want in my opinion. Basically like you've said.

Either way, this shouldn't be a use-case :)